src/http/client.cc

```diff
@@ -125,14 +125,17 @@ future<connection::reply_ptr> connection::recv_reply() {
     http_response_parser parser;
     return do_with(std::move(parser), [this] (auto& parser) {
         parser.init();
-        return _read_buf.consume(parser).then([&parser] {
+        return _read_buf.consume(parser).then([this, &parser] {
             if (parser.eof()) {
                 throw std::runtime_error("Invalid response");
             }
 
             auto resp = parser.get_parsed_response();
             sstring length_header = resp->get_header("Content-Length");
             resp->content_length = strtol(length_header.c_str(), nullptr, 10);
+            if ((resp->_version != "1.1") || seastar::internal::case_insensitive_cmp()(resp->get_header("Connection"), "close")) {
+                _persistent = false;
```

Comment:
if we are aiming at a standard compliant implementation, then we need to check for "Keep-Alive" in header. in a gist, the check would look like:
```c++
if (resp->_version == "1.1") {
    // stay open unless explicitly closed
    if (seastar::internal::case_insensitive_cmp()(resp->get_header("Connection"), "close")) {
        _persistent = false;
    }
} else {
    // for older HTTP, Keep-Alive must be explicitly negotiated
    _persistent = false;
    if (resp->get_header("Keep-Alive").empty()) {
        _persistent = true;
    } else if (seastar::internal::case_insensitive_cmp()(resp->get_header("Connection"), "keep-alive")) {
        _persistent = true;
    }
}
````
see also https://www.rfc-editor.org/rfc/rfc2068#section-19.7.1.1

Reply:
Are you sure about that? I didn't remember this header so I went and look at documentation, and it appears that this "Keep-Alive" header is just used as an unofficial (not defined in any RFC)  way to add parameters like timeout to the keepalive.

You suggested that the "Keep-alive:" header had special meaning in HTTP 1.0. Can you find a reference to that?

I think it was "Connection: Keep-Alive", not "Keep-Alive: ...", which mattered in HTTP 1.0. But it was so long ago my memory is hazy ;-)

Reply:
For older servers the keep-alive is set by the server _only_ if the client does so. Seastar client doesn't so there's no chance it will see keep-alive in 1.0 server response

Reply:
```
} else {
    // for older HTTP, Keep-Alive must be explicitly negotiated
    _persistent = false;
    if (resp->get_header("Keep-Alive").empty()) {
        _persistent = true;
    } else if (seastar::internal::case_insensitive_cmp()(resp->get_header("Connection"), "keep-alive")) {
        _persistent = true;
    }
}
```

:point_up: if keep-alive header is missing, why _persistent = true? It's older server that is not persistent by default

Reply:
> For older servers the keep-alive is set by the server _only_ if the client does so. Seastar client doesn't so there's no chance it will see keep-alive in 1.0 server response

This is true (we discussed this above). In theory it's best not to couple the sending part with the response-parsing part of the code, but it's fine that we do - if we ever want to change one to support keep-alive on http 1.0 (and we won't...) we'll just need to change the other.
In any case, none of this explains why @tchaikov asked to look at the "Keep-Alive" header. The so-called "Keep-Alive Connection token" mentioned in the RFC is the "Connection: Keep-Alive" header, not "Keep-Alive: ...".

Reply:
> Are you sure about that? 

i referenced the related section of RFC. does it explain the rationales?

> For older servers the keep-alive is set by the server only if the client does so. Seastar client doesn't so there's no chance it will see keep-alive in 1.0 server response

ahh, i see. so it's determined by the (faulty) client side. but there is a, well hypothesized use case, where Seastar is used to implement a proxy which proxies the traffic from an ancient client. and we are supposed to cache the connections, as that's what proxy does. and now that we do proxy, we might want to use the same HTTP protocol as the ancient client does. in other words, we do what the client does.

Reply:
> I think it was "Connection: Keep-Alive", not "Keep-Alive: ...", which mattered in HTTP 1.0. But it was so long ago my memory is hazy ;-)

okay, i think i misinterpreted the section.

> 
>    When the Keep-Alive connection-token has been transmitted with a
>    request or a response, a Keep-Alive header field MAY also be
>    included. The Keep-Alive header field takes the following form:

so we don't need to check it if we already check the one in "Connection".

Reply:
> okay, i think i misinterpreted the following paragraph:
> 
> > When the Keep-Alive connection-token has been transmitted with a
> > request or a response, a Keep-Alive header field MAY also be
> > included. The Keep-Alive header field takes the following form:
> 
> so we don't need to check it if we already check the one in "Connection".

Right. If you look more carefully (I agree the wording is confusing), the phrase "Keep-Alive connection-token" refers to "Connection: Keep-Alive". It's not related to the "Keep-Alive:" header which is documented in https://datatracker.ietf.org/doc/html/rfc2068#section-19.7.1.1 and has no standard semantics so I think there's absolutely nothing we need to do with this header. (only if we were a proxy, we should have taken care to copy it, or something).

Reply:
based on our discussion above, i revised the proposed check:
```c++
if (resp->_version == "1.1") {
    // stay open unless explicitly closed
    _persistent = !seastar::internal::case_insensitive_cmp()(resp->get_header("Connection"), "close");
} else {
    // for older HTTP, Keep-Alive must be explicitly negotiated
    _persistent = seastar::internal::case_insensitive_cmp()(resp->get_header("Connection"), "keep-alive");
}
````

Reply:
This was exactly my proposal above (https://github.com/scylladb/seastar/pull/1841#discussion_r1331404362).
But @xemul rightly noted that according to the protocol, the 1.0 server can never respond with "Connection: Keep-Alive" if the client didn't have the same header in its request, and since we know that we didn't (we're talking about the client code here), the server also can't. So in the 1.0 case, "_persistent = false" is correct in our case. At least until we change our code to send "Connection: Keep-Alive", which I'm pretty sure we never will because nobody cares about the performance on 1.0 servers (which don't exist any more).